runtime/parachains/src/inclusion_inherent.rs

```diff
@@ -127,6 +129,47 @@ decl_module! {
 	}
 }
 
+/// Limit the number of backed candidates processed in order to stay within block weight limits.
+///
+/// Use a configured assumption about the weight required to process a backed candidate and the
+/// current block weight as of the execution of this function to ensure that we don't overload
+/// the block with candidate processing.
+///
+/// If the backed candidates exceed the available block weight remaining, then skips all of them.
+/// This is somewhat less desirable than attempting to fit some of them, but is more fair in the
+/// even that we can't trust the provisioner to provide a fair / random ordering of candidates.
+fn limit_backed_candidates<T: Config>(
+	backed_candidates: Vec<BackedCandidate<T::Hash>>,
+) -> Vec<BackedCandidate<T::Hash>> {
+	const BACKED_CANDIDATE_WEIGHT_ASSUMPTION: usize = 10_000;
+
+	let block_weight_remaining = <T as frame_system::Config>::MaximumBlockWeight::get()
+		.saturating_sub(frame_system::Module::<T>::block_weight().total());
+
+	if backed_candidates.len() * BACKED_CANDIDATE_WEIGHT_ASSUMPTION > block_weight_remaining as usize {
```

Comment:
It seems weight is added 2 times here. I mean at the moment this is executed (i.e. the extrinsic is applied) frame_system block weight already include the extrinsic weight (it is added by check_weight extension). So actually if you want to check if `inclusion` has reached block weight limit you just need to check if here block weight is more than block weight limit.

But I see that `inclusion` weight is `1_000_000_000` and so doesn't include the weight part factor of `backed_candidates.len()`. But then you never write this amount into the used weight amount.

So either it is documented that `inclusion` weight doesn't include the full weight as it doesn't include the variable part factor of `backed_candidates.len()`. And we write this weight somewhere.

Or (and I think it is the best) we use refund here:
Basically we can have precise and full weight info for `inclusion` (so either benchmark or manual estimation but this will end up being a formula `K1 + backed_candidates.len() * K2`). and here we just check if we went beyond the full block weight.
In case we reached the block weight limit then we don't include any candidates.
At the end of inclusion we can still do a weight refund if no candidate were included (i.e. `K1`).

Reply:
3dbed6d9